static/js/filter.js

```diff
@@ -382,13 +382,48 @@ Filter.prototype = {
     },
 
     can_mark_messages_read: function () {
-        return _.every(this._operators, function (elem) {
-            return (_.contains(['stream', 'topic', 'pm-with'], elem.operator)
-                || elem.operator === 'is' && elem.operand === 'private'
-                || elem.operator === 'in' && elem.operand === 'all'
-                || elem.operator === 'in' && elem.operand === 'home')
-                && !elem.negated;
-        });
+        const term_types = this.sorted_term_types();
+
+        if (_.isEqual(term_types, ['stream', 'topic'])) {
+            return true;
+        }
+
+        if (_.isEqual(term_types, ['pm-with'])) {
+            return true;
+        }
+
+        // Folks like me really hate it when Zulip
+        // marks messages as read in interleaved views,
+        // so we will eventually have a setting that
+        // early-exits before the subsequent checks.
```

Comment:
Probably should be this, since "I" isn't ideal in code comments :).
```
// TODO: Some users really hate it when Zulip marks messages as read
// in interleaved views, so we will eventually have a setting
// that early-exits before the subsequent checks.
```